controllers/placementapi_controller.go

```diff
@@ -318,8 +321,73 @@ func (r *PlacementAPIReconciler) Reconcile(ctx context.Context, req ctrl.Request
 	return r.reconcileNormal(ctx, instance, helper)
 }
 
+// fields to index to reconcile when change
+const (
+	passwordSecretField     = ".spec.secret"
+	caBundleSecretNameField = ".spec.tls.caBundleSecretName"
+	tlsAPIInternalField     = ".spec.tls.api.internal.secretName"
+	tlsAPIPublicField       = ".spec.tls.api.public.secretName"
+)
+
+var (
+	allWatchFields = []string{
+		passwordSecretField,
+		caBundleSecretNameField,
+		tlsAPIInternalField,
+		tlsAPIPublicField,
+	}
+)
+
 // SetupWithManager sets up the controller with the Manager.
 func (r *PlacementAPIReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// index passwordSecretField
+	if err := mgr.GetFieldIndexer().IndexField(context.Background(), &placementv1.PlacementAPI{}, passwordSecretField, func(rawObj client.Object) []string {
+		// Extract the secret name from the spec, if one is provided
+		cr := rawObj.(*placementv1.PlacementAPI)
+		if cr.Spec.Secret == "" {
```

Comment:
I previously had a quick look to use something like `reflect.Indirect(v).FieldByName(field)` but it did not work so we considered that a later optimization work to look into. agreed to check this in a follow up

Reply:
ok so i agree this can  be refactored to reduce the duplications

the lamda/funciton passed to IndexField  is just returning a StringSlice/charArray or nil if the filed is not set.

we know the mapphng between the string constants and the struct fields so we just need a pointer to member data for each and store a map of "string" -> "pointer to member data" which we can apply the the spec object to select the data 

if this was c++ that would be simple in golang I'm not sure that this easlisly do able so a slight different approach is needed 
```
field_indexers := map[string]interface{}{
    passwordSecretField: func(rawObj client.Object) []string {
		// Extract the secret name from the spec, if one is provided
		cr := rawObj.(*placementv1.PlacementAPI)
		if cr.Spec.Secret == "" {
			return nil
		}
		return []string{cr.Spec.Secret}
	},
	caBundleSecretNameField: func(rawObj client.Object) []string {
		// Extract the secret name from the spec, if one is provided
		cr := rawObj.(*placementv1.PlacementAPI)
		if cr.Spec.TLS.CaBundleSecretName == "" {
			return nil
		}
		return []string{cr.Spec.TLS.CaBundleSecretName}
	},
	tlsAPIInternalField: func(rawObj client.Object) []string {
		// Extract the secret name from the spec, if one is provided
		cr := rawObj.(*placementv1.PlacementAPI)
		if cr.Spec.TLS.API.Internal.SecretName == nil {
			return nil
		}
		return []string{*cr.Spec.TLS.API.Internal.SecretName}
	},
	tlsAPIPublicField: func(rawObj client.Object) []string {
		// Extract the secret name from the spec, if one is provided
		cr := rawObj.(*placementv1.PlacementAPI)
		if cr.Spec.TLS.API.Public.SecretName == nil {
			return nil
		}
		return []string{*cr.Spec.TLS.API.Public.SecretName}
	}
}

for field, indexer := range field_indexers{
    if err := mgr.GetFieldIndexer().IndexField(context.Background(), &placementv1.PlacementAPI{}, field, indexer
);err != nil {
		return err
	}

}
```

i think that is the same and will work but its not much simpler

the functor objects/lamdas are basically just c++ pointer to data members

Reply:
i have an example of this patteren here

https://github.com/openstack-k8s-operators/nova-operator/blob/8f204945b887a562229143f40901c4540925e9cd/controllers/common.go#L294-L341